forem/forem
# Manual Audience Segments API write endpoints

## What type of PR is this? (check all applicable)


## Description

API endpoints (`create`, `destroy`, `add_users`, `remove_users`) that modify audience segments.

This is still a fairly large PR, but a good deal of it is Swagger docs

## Related Tickets & Documents



- Related Issue #19438 


## QA Instructions, Screenshots, Recordings

These endpoints should all only work with an API key that has admin access to `AudienceSegment`, and should fail if used with a non-manual audience segment ID (returning 404)

- `POST /api/segments` should create a manual audience segment and return it.
- `DELETE /api/segments/{id}` should delete the provided manual segment and return it
    - It should return a readable error if the segment is currently being used by a billboard
    - You can add a segment to a billboard with `DisplayAdd.find(billboard_id).update!(audience_segment: segment)`
- `PUT /api/segments/{id}/add_users` should add the users with the IDs passed in the request body (`{ user_ids: [<ids>] }`) to the provided manual segment.
    - It should return JSON containing `"succeeded` (an array of user IDs that were processed successfully and `"failed"` (user IDs that could not be added to the segment)
    - It should return a readable error if user IDs are not provided or if more than 10,000 are provided
- `PUT /api/segments/{id}/remove_users` should work the same as above, but removing users from the segment instead of adding them.

## Added/updated tests?


## [optional] Are there any post deployment tasks we need to perform?

I am not quite sure - there's a migration that introduces a new index, which might need to be tested on canary first
